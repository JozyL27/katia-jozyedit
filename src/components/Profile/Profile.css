.profile__icons-container {
    display: flex;
    justify-content: space-around;
    align-items: center;
    padding: 1rem 0;
    height: 9vh;
    max-height: 7rem; 
    width: 85%;
    margin: 0 auto .75rem;
}

.profile__edit-icon {
    height: 7vh;
    max-height: 5rem;
}

.profile__edit-icon:hover {
    cursor: pointer;
}

.profile__back-icon {
    height: 7vh;
    max-height: 5rem;
}

.profile__section {
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    border: .1rem solid lightgrey;
    max-width: 85%;
    margin: 0 auto;
    border-radius: 2.5vh;
    background-color: white;
    box-shadow: .5rem 1rem .8rem #888888;
}

.profile__section > * {
    font-size: 3.5vh;
}

.profile__img-name-div {
    width: calc(100% + .4rem);
    margin-top: -.5vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    border-radius: 2.5vh;
    background: #30CFD0;
    box-shadow: .2rem 1rem .8rem #888888;
}

.profile__img-container {
    max-width: 90%;
}

.profile__img {
    display: block;
    width: auto;
    height: 35vh;
    max-height: 28rem;
    border-radius: 2.5vh;
    margin: 1.5vh 0 0;
}

.profile__displayName {
    padding: 1vh;
    color: black;
    font-size: 3vh;
    font-weight: bold;
}

.profile__card-header {
    width: 100%;
    height: 5.5vh;
    text-align: left;
    padding: .7vh 2vh;
    font-size: 3.5vh;
    font-weight: bold;
    margin: 1vh 0;
    text-align: center;
    border: none;
    border-bottom: .3rem solid lightgray;
    width: 60%;
}

.profile__genres, .profile__gamer-tags {
    text-align: center;
}

.profile__bio {
    width: 90%;
    text-align: center;
    margin-bottom: 1vh;
}

.logoutLink-container {
    display: flex;
    justify-content: center;
    padding: 1.5vh 0;
    margin: 1vh;
}

.logoutLink-container > * {
    font-size: 3.5vh;
    margin: 0 auto;
}

.profile__editForm {
    display: flex;
    flex-direction: column;
    justify-content: center;
    width: 80%;
    margin: 0 auto;
}

.profile__editCancelSubmit-div {
    display: flex;
    justify-content: space-around;
    align-items: center;
    padding: 1vh 0;
    height: 9vh;
    max-height: 12.6rem; 
}

.profile__editCancel {
    height: 7vh;
    max-height: 8rem;
}

.profile__editCancel:hover {
    cursor: pointer;
}

.profile__editSubmit {
    height: 7vh;
    max-height: 8rem;
}

.profile__editSubmit:hover {
    cursor: pointer;
}

.profile__ImgEdit-container {
    border-radius: .8vh;
    padding: 2vh .5vh;
    margin: 0 auto 2rem;
    width: 100%;
}

.profile__ImgEdit {
    height: 25vh;
    width: auto;
    display: block;
    margin: 0 auto;
    border-radius: .8vh;
}

.profile__imageEditInput {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-evenly;
    width: 80%;
    margin: 2rem auto 0;
}

.profile__file-upload {
    font-size: 2.5vh;
    margin: 1rem 0 2rem;
}

.profile__imageEditInput > .profile__ImgEdit-submit {
    font-size: 2vh;
    padding: .6rem 1.5rem;
}

.profile__editForm > label, .profile__gamer-tags-header {
    padding: .2vh 1.5vh;
    font-size: 3vh;
    font-weight: bold;
    height: 4.5vh;
    text-align: center;
    width: 70%;
    margin: 3.5rem auto;
    color: black;
    border: none;
    border-bottom: .2rem solid lightgrey;
}

#display-name {
    font-size: 3vh;
    padding: .5vh 0 0 1.5vh;
    height: 5.5vh;
    border: .08rem solid gray;
    width: 80%;
    margin: 0 auto;
    text-align: center;
    border-radius: .8vh;
}

.profile__editForm > textarea {
    resize: none;
}

#lfm {
    font-size: 3vh;
    padding: .5vh 0 0 1.5vh;
    height: 12vh;
    border: .08rem solid gray;
    width: 100%;
    margin: 0 auto;
    border-radius: .8vh;
}

#platforms {
    display: flex;
    justify-content: space-around;
    margin: 0;
    padding: 1.4vh 0;
    width: 100%;
    margin: 0 auto;
}

#genres {
    text-align: center;
    height: 4.5vh;
    font-size: 3vh;
    padding-left: 1.5vh;
}

#bio {
    font-size: 3vh;
    padding: .5vh 0 0 1.5vh;
    height: 20vh;
    border: .08rem solid gray;
    max-width: 100%;
    margin: 0rem auto 4vh auto;
    border-radius: .8vh;
}

.profile__editGenreList {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    justify-content: center;
    width: 100%;
    margin: 0 auto;
}

.profile__editGenre {
    display: inline-flex;
    align-items: center;
    white-space: nowrap;
    background-color: #330867;
    color: white;
    border-radius: .5vh;
    padding: .5vh 0 .35vh .5vh;
    margin: .75vh;
}

.profile__editGenreLabel {
    font-size: 2.6vh;
    font-weight: 600;
}

.profile__editGenreButton {
    display: inline-block;
    margin: 0 .75vh 0 .6vh;
    padding-bottom: .2vh;
    height: auto;
    width: 1.8vh;
}

.profile__editGenre:hover {
    background-color: #d0d0d0;
}

.profile__editGenre:hover > .profile__editGenreButton {
    stroke: rgb(224, 0, 0);
    width: 2.1vh;
    margin: 0 .6vh 0 .45vh;
    stroke-width: 2.5;
}

.profile__genreSelect {
    font-size: 2.6vh;
    font-weight: 600;
    background-color: rgb(19, 93, 177);
    border-radius: .5vh;
    border: .1rem solid rgb(19, 93, 177);
    padding: .5vh 0 .65vh .35vh;
    margin: .75vh .2rem;
    text-align: center;
    color: white;
}

.profile__genreSelectItem {
    border: .1rem solid #e0e0e0;
    font-size: 2.6vh;
    font-weight: 600;
}

.imageError {
    font-weight: bold;
    color: rgb(182, 21, 21);
    text-align: center;
    padding: 10px;
    margin-top: 10px;
    font-size: 2.5vmax;
}

@media screen and (min-width: 600px) {
    .profile__icons-container {
        max-width: 70%;
    }

    .profile__section {
        max-width: 70%;
    }

    .profile__section > * {
        font-size: 3vh;
    }
}

@media screen and (min-width: 800px) {
    .profile__editForm {
        width: 60%;
    }
}

@media screen and (min-width: 900px) {
    .profile__icons-container {
        max-width: 55%;
    }

    .profile__section {
        max-width: 615px;
    }

    .profile__section > * {
        font-size: 2.5vh;
    }
}

@media screen and (min-width: 950px) {
    .profile__editForm {
        width: 50%
    }
}