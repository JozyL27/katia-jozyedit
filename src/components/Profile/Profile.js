import React from 'react'
import './Profile.css'
import { Link } from 'react-router-dom'
import editIcon from '../../images/edit-icon.svg'
import nintendoLogo from '../../images/nintendo_logo.png'
import nintendoLogoGray from '../../images/nintendo_logo_b&w.png'
import playstationLogo from '../../images/playstation_color_png.png'
import playstationLogoGray from '../../images/playstation_b&w_png.png'
import PC_Logo from '../../images/PC_Keyboard_Mouse_Icon.png'
import PC_LogoGray from '../../images/PC_Keyboard_Mouse_Icon_b&w.png'
import xboxLogo from '../../images/xbox_logo_png.png'
import xboxLogoGray from '../../images/xbox_logo_b&w_png.png'
import checkmarkSVG from '../../images/checkmark.svg'
import xSVG from '../../images/x.svg'
import UserContext from '../../Contexts/UserContext'
import ProfileService from '../../services/profile-service'
import GamerTag from '../GamerTag/GamerTag'
import rightArrow from '../../images/right-arrow.svg'
import ConfirmDialog from '../Confirm/ConfirmDialog';
import customConfirm from '../../Utils/CustomConfirm';

export default class Profile extends React.Component {
    static contextType = UserContext

    state = {
        isEditing: false,
        selectedFile: null,
        avatar: '',
        display_name: '',
        lfm_in: '',
        bio: '',
        platforms: [],
        genres: [],
        allGenres: [],
        currGenre: '',
        gamer_tags: {
            xbox: '',
            psn: '',
            nintendo: '',
            steam: '',
            discord: '',
            other: ''
        },
        error: null,
        loading: true,
    }

    componentDidMount() {
        ProfileService.getAllUserGenres()
        .then(res => this.setState({ allGenres: res }))
        .catch(error => this.setState({error: error.message}))

        ProfileService.getProfile(this.context.user_id)
        .then(user => this.setState({ 
            avatar: user.avatar,
            display_name: user.display_name,
            lfm_in: user.lfm_in,
            bio: user.bio,
            platforms: user.platforms,
            genres: user.genres,
            gamer_tags: {
              xbox: user.xbox,
              psn: user.psn,
              nintendo: user.nintendo,
              steam: user.steam,
              discord: user.discord,
              other: user.other
            },
            loading: false
         }))
         .catch(error => this.setState({error: error.message}))
    }

    handleEditButton = event => {
        event.preventDefault()
        this.setState({ isEditing: true })
    }

    saveEdit = event => {
        event.preventDefault()
        const { user_id } = this.context
        const userInfo = {
            display_name: this.state.display_name,
            lfm_in: this.state.lfm_in,
            bio: this.state.bio,
            platforms: this.state.platforms,
            genres: this.state.genres,
            xbox: this.state.gamer_tags.xbox,
            psn: this.state.gamer_tags.psn,
            nintendo: this.state.gamer_tags.nintendo,
            steam: this.state.gamer_tags.steam,
            discord: this.state.gamer_tags.discord,
            other: this.state.gamer_tags.other
        }
        
        customConfirm(ConfirmDialog, 'Save Profile').then(answer => {
            if(answer) {
                ProfileService.updateProfile(user_id, userInfo)
                .then(user => {
                    this.setState({ 
                        isEditing: false,
                        display_name: user.display_name,
                        lfm_in: user.lfm_in,
                        bio: user.bio,
                        platforms: user.platforms,
                        genres: user.genres,
                        gamer_tags: {
                            xbox: user.xbox,
                            psn: user.psn,
                            nintendo: user.nintendo,
                            steam: user.steam,
                            discord: user.discord,
                            other: user.other
                          }
                    })
                })
                .catch(error => this.setState({error: error.message}))
            }
        })
    }

    cancelEdit = () => {
       customConfirm(ConfirmDialog, 'Cancel Edit').then(answer => {
           if(answer) {
                ProfileService.getAllUserGenres()
                .then(res => this.setState({ allGenres: res }))
                .catch(error => this.setState({error: error.message}))

                ProfileService.getProfile(this.context.user_id)
                .then(user => this.setState({ 
                    avatar: user.avatar,
                    display_name: user.display_name,
                    lfm_in: user.lfm_in,
                    bio: user.bio,
                    platforms: user.platforms,
                    genres: user.genres,
                    isEditing: false,
                    selectedFile: null,
                }))
                .catch(error => this.setState({error: error.message}))
            }
       })
    }

    handleLogoutClick = () => {
        this.context.processLogout()
    }

    handleDisplayNameChange = event => {
        this.setState({ display_name: event.target.value })
    }

    handleLookingForChange = event => {
        this.setState({ lfm_in: event.target.value })
    }

    handleBioChange = event => {
        this.setState({ bio: event.target.value })
    }

    handleGamerTagsChange = event => {
        const platform = event.target.name
        const text = event.target.value
        // initialize the value of our state
        const newValue = {
            xbox: this.state.gamer_tags.xbox,
            psn: this.state.gamer_tags.psn,
            nintendo: this.state.gamer_tags.nintendo,
            steam: this.state.gamer_tags.steam,
            discord: this.state.gamer_tags.discord,
            other: this.state.gamer_tags.other
        }
        newValue[platform] = text
        this.setState({ gamer_tags: 
            newValue
        })
    }

    handleGenresChange = event => {
        this.setState({ genres: event.target.value })
    }

    handlePlatformsChange = event => {
        const platform = event.target.name
        this.state.platforms.includes(platform) ?
            this.setState({
                platforms: this.state.platforms.filter(item => item !== platform)
            })
            :
            this.setState({
                platforms: [platform, ...this.state.platforms]
            })
    }

    generateEditPlatforms = () => {
        return (
            <>
            {this.state.platforms.includes('Xbox') ? <img name='Xbox' className='main__xbox' src={xboxLogo} alt='Xbox logo' onClick={this.handlePlatformsChange}/> : <img  name='Xbox' className='main__xbox' src={xboxLogoGray} alt='Xbox logo gray' onClick={this.handlePlatformsChange} />}
            {this.state.platforms.includes('PlayStation') ? <img name='PlayStation' className='main__playstation' src={playstationLogo} alt='Playstation logo' onClick={this.handlePlatformsChange} /> : <img name='PlayStation'  className='main__playstation' src={playstationLogoGray} alt='Playstation logo gray' onClick={this.handlePlatformsChange} />}
            {this.state.platforms.includes('Nintendo') ? <img name='Nintendo' className='main__nintendo' src={nintendoLogo} alt='Nintendo logo' onClick={this.handlePlatformsChange} /> : <img name='Nintendo' className='main__nintendo' src={nintendoLogoGray} alt='Nintendo logo gray' onClick={this.handlePlatformsChange} />}
            {this.state.platforms.includes('PC') ? <img name='PC' className='main__PC' src={PC_Logo} alt='PC logo' onClick={this.handlePlatformsChange} /> : <img name='PC' className='main__PC' src={PC_LogoGray} alt='PC logo gray' onClick={this.handlePlatformsChange} />}
            </>
        )
    }
        
    handleSelectGenre = event => {
        this.setState({ genres: [...this.state.genres, event.target.value] })
    }

    avatarChangedHandler = event => {
        this.setState({
            selectedFile: event.target.files[0]
        });
    }

    avatarUploadHandler = event => {
        event.preventDefault();
        const formData = new FormData();
        formData.append('profileImg', this.state.selectedFile);
        ProfileService.uploadAvatar(this.context.user_id, formData)
            .then(res => {
                this.setState({ avatar: res.location });
            })
            .catch(error => this.setState({ error: error.error }));
    }

    genreToDelete = event => {
        event.preventDefault()

        let newGenres = this.state.genres.filter(item => item !== event.currentTarget.id)
        this.setState({ genres: newGenres })
    }

    render() {
        let allGenres = this.state.allGenres || []
        let userGenres = this.state.genres || []
        const { avatar, display_name, bio, lfm_in, genres, platforms, gamer_tags, loading } = this.state
        if (loading) {
            return (
                <div className="lds-roller"><div></div><div></div>
                <div></div><div></div><div></div><div>
                </div><div></div><div></div></div>
            )
        } else if(!this.state.isEditing) { // Checks isEditing value to conditionally render normal or editing page
            return (
                <>
                <section className='profile__icons-container'>
                    <img className='profile__edit-icon' onClick={this.handleEditButton} src={editIcon} alt='edit-icon' />
                    <Link to='/swipe'>
                        <img className='profile__back-icon' src={rightArrow} alt='swipe-back-icon' />
                    </Link>
                </section>
                <section className='profile__section'>
                    <div className='profile__img-name-div'>
                        <div className='profile__img-container'>
                            <img src={avatar} 
                            alt='avatar' className='profile__img' />
                        </div>
                        <span className='profile__displayName'>{display_name}</span>
                    </div>
                    <h4 className='profile__card-header'>Platforms</h4>
                    <div id='platforms'>
                        {platforms && platforms.includes("Xbox") ? <img className='main__xbox' src={xboxLogo} alt='Xbox logo' /> : null}
                        {platforms && platforms.includes("PlayStation") ? <img className='main__playstation' src={playstationLogo} alt='Playstation logo' /> : null}
                        {platforms && platforms.includes("Nintendo") ? <img className='main__nintendo' src={nintendoLogo} alt='Nintendo logo' /> : null}
                        {platforms && platforms.includes("PC") ? <img className='main__PC' src={PC_Logo} alt='PC logo' /> : null}
                    </div>
                    <h4 className='profile__card-header'>LFM</h4>
                    {this.context.generateLfmElements(lfm_in)}
                    <h4 className='profile__card-header'>Genres</h4>                    
                    <span className='profile__genres'>{this.context.generateGenreString(genres)}</span>
                    <h4 className='profile__card-header'>Gamer Tags</h4>
                    <div className='profile__gamer_tags-container'>
                        {gamer_tags.xbox !== '' ? <p className='profile__gamer-tags'>Xbox: {gamer_tags.xbox}</p> : null}
                        {gamer_tags.psn !== '' ? <p className='profile__gamer-tags'>PSN: {gamer_tags.psn}</p> : null}
                        {gamer_tags.nintendo !== '' ? <p className='profile__gamer-tags'>Nintendo: {gamer_tags.nintendo}</p> : null}
                        {gamer_tags.steam !== '' ? <p className='profile__gamer-tags'>Steam: {gamer_tags.steam}</p> : null}
                        {gamer_tags.discord !== '' ? <p className='profile__gamer-tags'>Discord: {gamer_tags.discord}</p> : null}
                        {gamer_tags.other !== '' ? <p className='profile__gamer-tags'>Other: {gamer_tags.other}</p> : null}
                    </div>
                    <h4 className='profile__card-header'>Bio</h4>
                    <span className='profile__bio'>{bio}</span>
                </section>
                <section className='logoutLink-container'>
                    <Link onClick={this.handleLogoutClick} 
                    to='/login' className='blue-button'>
                    Logout
                    </Link>
                </section>
                </>
            )
        } else if(this.state.isEditing) {
            return (
                <>
                <form className='profile__editForm' name='editForm' onSubmit={this.saveEdit}>
                    <div className='profile__editCancelSubmit-div'>
                        <img className='profile__editCancel' src={xSVG} alt='cancel-button' onClick={this.cancelEdit} />
                        <img className='profile__editSubmit' src={checkmarkSVG} alt='submit-button' onClick={this.saveEdit}/>
                    </div>
                    <section className='profile__ImgEdit-container'>
                        <img src={avatar} 
                        alt='avatar' className='profile__ImgEdit' />
                        <div className='profile__imageEditInput'>
                        <input type='file' className='profile__file-upload' onChange={this.avatarChangedHandler} />
                        {this.state.selectedFile && <button className='blue-button profile__ImgEdit-submit' 
                        onClick={this.avatarUploadHandler}>Upload</button>}
                        {this.state.error && <p className='imageError'> Error: {this.state.error}</p>}
                        </div>
                    </section>
                    <label htmlFor='display-name'>Display Name</label>
                    <input type='text' name='display-name' onChange={this.handleDisplayNameChange}
                    id='display-name' defaultValue={display_name} />
                    <label htmlFor='lfm'>LFM</label>
                    <textarea rows='7' cols='40' name='lfm' onChange={this.handleLookingForChange}
                    id='lfm' defaultValue={lfm_in} />
                    <label htmlFor='platforms'> Platforms</label>
                    <div type='text' name='platforms' 
                    id='platforms'>
                        {this.generateEditPlatforms()}
                    </div>
                    <label htmlFor='genres'>Genres</label>
                        <div className='profile__editGenreList'>
                            {userGenres.map((item, index) => 
                                <span key={index} className='profile__editGenre' id={item} onClick={(item) => this.genreToDelete(item)} aria-live='polite'>
                                    <label className='profile__editGenreLabel'>{item}</label>
                                    <svg className='profile__editGenreButton' xmlns="http://www.w3.org/2000/svg" width="22" height="22" viewBox="0 0 22 22" fill="none" stroke="#f00" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round">
                                        <line x1="20" y1="2" x2="2" y2="20"/>
                                        <line x1="2" y1="2" x2="20" y2="20"/>
                                    </svg>
                                </span>
                            )}
                            <select className='profile__genreSelect' onChange={this.handleSelectGenre} aria-live='polite'>
                                <option value='' style={{display: "none"}}>Choose Genre</option>
                                {allGenres.map((item, index) => 
                                userGenres.includes(item.genre) ? 
                                null 
                                : <option className='profile__genreSelectItem' key={index} value={item.genre}>{item.genre}</option>)}
                            </select>
                        </div>
                        
                    <h4 className='profile__gamer-tags-header'>Gamer Tags</h4>
                    {Object.entries(gamer_tags).map((item, idx) => {
                        return <GamerTag 
                            key={idx}
                            name={item[0]}
                            defaultVal={item[1]}
                            handleChange={this.handleGamerTagsChange}
                            />
                    })}
                    <label htmlFor='bio'>Bio (Max 250 chars.)</label>
                    <textarea rows='7' cols='40' name='bio' onChange={this.handleBioChange}
                    id='bio' defaultValue={bio} />
                </form>
                </>
            )
        }
    }
}