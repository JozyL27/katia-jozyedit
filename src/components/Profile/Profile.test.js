import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import Profile from './Profile'
import { UserProvider } from '../../Contexts/UserContext'

describe('<Profile />', () => {
  // Smoke test
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <MemoryRouter>
        <UserProvider>
          <Profile />
        </UserProvider>
      </MemoryRouter>, div)
    ReactDOM.unmountComponentAtNode(div)
  })
  // Snapshot test
  it('renders the UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <UserProvider>
            <Profile />
          </UserProvider>
        </MemoryRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})