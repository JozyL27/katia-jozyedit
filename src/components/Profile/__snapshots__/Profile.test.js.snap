// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> renders the UI as expected 1`] = `
<div
  className="lds-roller"
>
  <div />
  <div />
  <div />
  <div />
  <div />
  <div />
  <div />
  <div />
</div>
`;
